src/za/ac/sun/cs/green/service/simplify/ConstantPropogation.java

```diff
@@ -0,0 +1,133 @@
+import java.util.Map;
+import java.util.Set;
+import java.util.SortedMap;
+import java.util.SortedSet;
+import java.util.Stack;
+import java.util.TreeMap;
+import java.util.TreeSet;
+import java.util.logging.Level;
+
+import za.ac.sun.cs.green.Instance;
+import za.ac.sun.cs.green.Green;
+import za.ac.sun.cs.green.expr.Expression;
+import za.ac.sun.cs.green.service.BasicService;
+import za.ac.sun.cs.green.util.Reporter;
+import za.ac.sun.cs.green.expr.Constant;
+import za.ac.sun.cs.green.expr.IntConstant;
+import za.ac.sun.cs.green.expr.IntVariable;
+import za.ac.sun.cs.green.expr.Operation;
+import za.ac.sun.cs.green.expr.Variable;
+import za.ac.sun.cs.green.expr.Visitor;
+import za.ac.sun.cs.green.expr.VisitorException;
+
+public class ConstantPropogation extends BasicService {
+	
+	private int invocations = 0;
+
+	public ConstantPropogation(Green solver) {
+		super(solver);
+	}
+
+	@Override
+    public Set<Instance> processRequest(Instance instance) {
+        @SuppressWarnings("unchecked")
+        Set<Instance> result = (Set<Instance>) instance.getData(getClass());
+        if (result == null) {
+            final Map<Variable, Variable> map = new HashMap<Variable, Variable>();
+            final Expression e = simplify(instance.getFullExpression(), map);
+            final Instance i = new Instance(getSolver(), instance.getSource(), null, e);
+            result = Collections.singleton(i);
+            instance.setData(getClass(), result);
+        }
+        return result;
+    }
+
+    @Override
+    public void report(Reporter reporter) {
+        reporter.report(getClass().getSimpleName(), "invocations = " + invocations);
+    }
+
+	public Expression simplify(Expression expression, Map<Variable, Variable> map) {
+		try {
+			log.log(Level.FINEST, "Before Simplification: " + expression);
+			invocations++;
+			SimplifyVisitor simplifyVisitor = new SimplifyVisitor();
+			expression.accept(simplifyVisitor);
+			expression = simplifyVisitor.getExpression();
+			log.log(Level.FINEST, "After simplification: " + expression);
+		} catch (VisitorException x) {
+			log.log(Level.SEVERE, "encountered an exception -- this should not be happening!", x);
+		} finally {
+			return expression;
+		}		
+	}
+
+	public class SimplifyVisitor extends Visitor {
+		private Stack<Expression> stack;
+		private Map<IntVariable, IntConstant> map;
+
+		public SimplifyVisitor() {
+			stack = new Stack<Expression>();
+			map = new TreeMap<IntVariable, IntConstant>();
+		}
+
+		public Expression getExpression() {
+			return stack.pop();
+		}
+
+		@Override
+		public void postVisit(IntConstant constant) {
+			stack.push(constant);
+		}
+
+		@Override
+		public void preVisit(Operation operation) {
+			Operation.Operator op = operation.getOperator();
+			if (op.equals(Operation.Operator.EQ)) {
+				Expression opL = operation.getOperand(0);
+				Expression opR = operation.getOperand(1);
+				if ((opL instanceof IntConstant) && (opR instanceof IntVariable)) {
```

Comment:
Adding a comment or two in here will help with the "friendliness factor" of this piece of code. I can see what you are doing, but never underestimate the power of some Javadoc :+1: